UML-Diagram/SearchSystem.puml

```diff
@@ -1,18 +1,70 @@
 @startuml SearchSystem
 package "Search System" {
+
+    ' Define abstract classes
+
+    abstract class SearchBase {
+
+        - Manager : M
+        + PerformSearch()
+        + Search()
+        + SearchResult()
+        + DisplayDetails()
+        + NoResultsFound()
+
+    }
+
+    abstract class PatronSearch extends SearchBase {
+        - PatronManager Patrons
+        + override DisplayDetails()
+        + override NotResultFound()
+        + abstract Search()
+    }
+
+    abstract class BookSearch extends SearchBase {
+        - BookManager Books
+        + override DisplayDetails()
+        + override NotResultFound()
+        + abstract Search()
+    }
     
     ' Define main classes
-    class BookSearch {
-        - BookManager : BooksList
-        + searchBookByTitle()
-        + searchBookByAuthor()
-        + searchBookByISBN()
+    class PatronSearchByName extends PatronSearch{
+        - PatronManager Patrons
+        + override Search()
+    }
+
+    class PatronSearchByMembership extends PatronSearch{
+        - PatronManager Patrons
+        + override Search()
+    }
+
+    class BookSearchByName extends BookSearch {
+        - BookManager : Books
+        + override Search()
+    }
+
+    class BookSearchByTitle extends BookSearch {
+        - BookManager : Books
+        + override Search()
+    }
+
+    class BookSearchByISBN extends BookSearch {
+        - BookManager : Books
+        + override Search()
+    }
+    
+    class BookSearchByAuthor extends BookSearch {
+        - BookManager : Books
+        + override Search()
+    }
+
+    class PerformPatronSearch {
+        + PatronPerformSearch() 
     }
 
-    class PatronSearch {
-        - PatronManager : PatronsList
-        + searchPatronByName()
-        + searchPatronByMembershipNumber()
+     class PerformBookSearch {
+        + BookPerformSearch() 
```

Comment:
It would be better to implement a common interface (SearchStrategy) for all search strategies.

Reply:
Yes, but there is the SearchBase that its the template for search strategies, but that its a base after that im using a specialization that would be the BookSearch and the PatronSearch, both of them implements searchbase but also override some methods like display.

Reply:
Nice